### PR TITLE
Alertmanager: Unregister tenant-specific metrics when the configuration is removed 

### PR DESCRIPTION
**Pull request by gotjosh**
**What this PR does**:

Unregisters tenant-specific metrics for the Alertmanager when the configuration is removed.

**Checklist**
- [x] Tests updated
- N/A Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by gouthamve** [comment]
Why are we still doing active and non-active? Can we just remove the tenant from memory?

**Comment by gotjosh** [comment]
@gouthamve Do we need that now? That'll probably be a more involved refactor and would like to avoid it, if possible. With the new proposal, I'm about to submit this is all going away anyways.

File: pkg/alertmanager/alertmanager_metrics.go
Comment (pstibrany) [inline pkg/alertmanager/alertmanager_metrics.go:160]: This only works properly if no aggregate metrics are reported. See `TestMetricsArePerUser` test in ruler for how this can be verified, or wait until https://github.com/cortexproject/cortex/pull/3511 lands for "soft" deletion of metrics, if you need aggregates (doesn't seem like it though).

File: pkg/alertmanager/multitenant.go
Comment (jtlisi) [inline pkg/alertmanager/multitenant.go:445]: **suggestion(if-minor):**
```suggestion

		// If we went from paused to active, we need to re-add the user-regitry.
		if !activeBefore && existings.IsActive() {
			am.alertmanagerMetrics.addUserRegistry(cfg.User, existing.registry)
		}
```

File: pkg/alertmanager/alertmanager_metrics.go
Comment (jtlisi) [reply pkg/alertmanager/alertmanager_metrics.go:160]: The alertmanager aggregates a significant number of its metrics so this PR is blocked by #3511

File: CHANGELOG.md
Comment (jtlisi) [inline CHANGELOG.md:6]: **suggestion(non-blocking):** 
```suggestion
* [CHANGE] Alertmanager: Per-tenant metrics will be unregistered when the alertmanager configuration for a tenant is removed. #3507
```

`s/Tenant-specific/Per-tenant/g`

File: pkg/alertmanager/alertmanager_metrics.go
Comment (pracucci) [reply pkg/alertmanager/alertmanager_metrics.go:160]: Given #3511 has been merged, is there any reason why this PR is not using the new `util.UserRegistries` and 9 RemoveUserRegistry()`?

Comment (jtlisi) [reply pkg/alertmanager/alertmanager_metrics.go]: IMO, It would make sense to use soft deletes now that #3511 is merged.

Comment (gotjosh) [reply pkg/alertmanager/alertmanager_metrics.go]: I actually went for soft as well.

**Review by jtlisi — APPROVED** [review]
LGTM

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:9]: We've just cut `1.6.0-rc.0`. Could you rebase `master` and move the CHANGELOG entry to the top, under the "master / unreleased" section, please?

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:448]: If I'm not missing anything, this should not be correct. The problem is that `existing.registry` is the previous Prometheus registry and not a new one. This means that the registered metrics are there since before `deleteUserRegistry()` has been called and this would lead to counters/histograms/summaries to be double tracked as soon as you re-add it via `addUserRegistry()`. Why? Because when we call `deleteUserRegistry()` the last value of all counters/histograms/summaries is "recorded" in the `UserRegistries`. When you add back the same registry via `addUserRegistry()`, the metrics in the registry will be summed up to the last value of the metrics when `deleteUserRegistry()` was called, but if the registry is the same you will end up with double countings.

Assuming we fix this issue, the only benefit I see of this PR is let **gauge metrics** values (only gauges) disappear from aggregated metrics when `deleteUserRegistry()`. If this is the desiderata of this PR, then to fix it we should use a new registry each time we call `addUserRegistry()`.

_Am I missing anything?_

Comment (gotjosh) [reply pkg/alertmanager/multitenant.go]: Thanks, Marco for pointing this out - I think you're right on the soft deletion part. We need new registry.

My intention was to remove _all_ metrics, but under the current circumstances (where we can't distinguish between a user moving away from a shard or being removed completely from the system) if this is the best we can do (without incurring in more complexity) - I'll take it.

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:448]: I can't understand this change honestly. You're just setting a new prometheus registry to `alertmanagerMetrics` but that registry will never get alertmanager metrics registered on.

I still believe this PR has not been conceived correctly. What I think you should do is completely release the alertmanager resources once a tenant is removed from this replica, and create a new alertmanager (with a new registry which is then registered to `alertmanagerMetrics`) when the tenant is re-added back. Basically this should completely remove the `IsActive()` logic.

Comment (gotjosh) [reply pkg/alertmanager/multitenant.go:448]: > I can't understand this change honestly. You're just setting a new prometheus registry to alertmanagerMetrics but that registry will never get alertmanager metrics registered on.

Sorry Marco, I completely left this PR in an abandoned state but my ultimate intention was to close it. I quickly realised exactly what you pointed out that the new registry is never initialised with the corresponding metrics.

Given I'm doing work around the area with the scalability, and completely removing the `IsActive` is a bit more involved than I initially expect, I'll close this and defer to later.

**Pull request closed without merging**
